fms/training/plugins.py

```diff
@@ -0,0 +1,228 @@
+
+
+class TrainerPlugin:
+    """
+    A TrainerPlugin runs once every epoch, and possibly every `steps` steps.
+    It is passed relevant objects that can be used for checkpointing, logging,
+    or validation.
+    """
+
+    def __init__(self, steps=None):
+        self.steps = steps
+
+    def run(self, step):
+        """
+        Whether or not to run this plugin on the current step.
+        """
+        # if step is None, we're at an epoch end not an intermediate step.
+        # By default we always run for epoch ends.
+        if step is None:
+            return True
+        # If self.steps is None, we're only recording epoch ends and this isn't one.
+        if self.steps is None:
+            return False
+        # record every `step` steps
+        if (step + 1) % self.steps == 0:
+            return True
+        return False
+
+    @abstractmethod
+    def step(
+        self,
+        model: nn.Module,
+        optimizer,
+        epoch: int,
+        metrics: Dict = {},
+        step: int = None,
+    ):
+        pass
+
+
+class InferenceValidator(TrainerPlugin):
+    """
+    A training plugin to print the results of running inference on a given prompt.
+    """
+
+    def __init__(
+        self, prompt_tokens: List[str], tokenizer, device, steps=None, eos_token=None
+    ):
+        super().__init__(steps)
+        self.tokenizer = tokenizer
+        input_ids = tokenizer.convert_tokens_to_ids(prompt_tokens)
+        self.input_ids = torch.tensor(input_ids, dtype=torch.long, device=device)
+        self.eos_token_id = (
+            None
+            if eos_token is None
+            else self.tokenizer.convert_tokens_to_ids([eos_token])[0]
+        )
+
+    def step(
+        self,
+        model: nn.Module,
+        optimizer,
+        epoch: int,
+        metrics: Dict = {},
+        step: int = None,
+    ):
+        if not self.run(step):
+            return
+        training = model.training
+        model.eval()
+        with torch.no_grad():
+            curtime = datetime.now().strftime("%H:%M.%S")
+            prefix = f"{curtime}:{epoch:02d}"
+            if step is not None:
+                prefix = prefix + f":{step:04d}"
+
+            result = generation.generate(model, self.input_ids, use_cache=True)
+            result = generation.truncate_after_eos(result, self.eos_token_id)
+            result = self.tokenizer.convert_ids_to_tokens(result)
+            result = self.tokenizer.convert_tokens_to_string(result)
+            print0("generated result:")
+            print0(result)
+        model.train(training)
+
+
+class MetricReporter(TrainerPlugin):
+    """
+    A training plugin to periodically log metrics
+    """
+
+    # TODO: add optional validation dataloader and validation loss.
+    # TODO: add `writer` functions that handles logging metrics to experiment
+    # tracking tools such as aimstack/wandb/neptune (or add alternate plugin?)
+    def __init__(self, seconds=10, writer=print0):
```

Comment:
can you add docstrings for this

Reply:
oh, I had doc strings at the class level. do you prefer them to be on the constructor itself?

Reply:
I added some more detail in the existing doc string, let me know if you prefer to move it.